src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2385 @@
+        
+        particle->mDeltaColor[i] = (end - start) / particle->mTimeToLive;
+    }
+    
+    // partice size
+    start = this->mStartSize + this->mStartSizeVariance * USFloat::Rand(-1.0f, 1.0f);
+    start = MAX(0, start);
+    
+    particle->mParticleSize = start;
+    
+    end = this->mFinishParticleSize + this->mFinishParticleSizeVariance * USFloat::Rand(-1.0f, 1.0f);
+    particle->mDeltaParticleSize = (end - start) / particle->mTimeToLive;
+    
+    
+    // particle rotation
+    start = this->mRotStart + this->mRotStartVariance * USFloat::Rand(-1.0f, 1.0f);
+    end = this->mRotEnd + this->mRotEndVariance * USFloat::Rand(-1.0f, 1.0f);
+    
+    particle->mParticleRotation = start;
+    particle->mDeltaParticleRotation = (end - start) / particle->mTimeToLive;
+    
+    // position
+    
+    // relative particle position
+    USVec3D loc;
+    if (this->mParticlePositionType == PARTICLE_POSITION_RELATIVE) {
+        loc = this->GetLoc();
+        particle->mStartPosition.Init(loc.mX, loc.mY);
+    }
+    // free particle position
+    else if (this->mParticlePositionType == PARTICLE_POSITION_FREE) {
+        
+        loc.Init(0.0f, 0.0f, 0.0f);
+        USAffine3D modelToWorld = this->GetLocalToWorldMtx ();
+        modelToWorld.Transform(loc);
+        particle->mStartPosition.Init(loc.mX, loc.mY);
+    }
+    
+    // set current position
+    particle->mCurrentPosition.Init(particle->mStartPosition);
+    
+    // direction
+    float a = D2R * (this->mAngle + this->mAngleVariance * USFloat::Rand(-1.0f, 1.0f));
+    
+    
+    // gravity mode
+    if ( this->mEmitterType == EMITTER_GRAVITY ) {
+        USVec2D v;
+        v.Init(Cos(a), Sin(a));
+        
+        float s = this->mSpeed + this->mSpeedVariance * USFloat::Rand(-1.0f, 1.0f);
+        v.Scale( s );
+        
+        // direction
+        particle->mDirection.Init(v);
+        
+        // gravity variance
+        particle->mGravity.mX = this->mGravity[0] + this->mGravityVariance[0] * USFloat::Rand(-1.0f, 1.0f);
+        particle->mGravity.mY = this->mGravity[1] + this->mGravityVariance[1] * USFloat::Rand(-1.0f, 1.0f);
+        
+        // radial accel
+        particle->mRadialAcceleration = this->mRadialAcceleration + this->mRadialAccelVariance * USFloat::Rand(-1.0f, 1.0f);
+        
+        // tangential accel
+        particle->mTangentialAcceleration = this->mTangentialAcceleration + this->mTangentialAccelVariance * USFloat::Rand(-1.0f, 1.0f);
+        
+    }
+    // radial mode
+    else {
+        float startRadius = this->mEndRadius + this->mEndRadiusVariance *  USFloat::Rand(-1.0f, 1.0f);
+        float endRadius = this->mStartRadius + this->mStartRadiusVariance *  USFloat::Rand(-1.0f, 1.0f);
+        
+        particle->mRadius = startRadius;
+        
+        particle->mDeltaRadius = (endRadius - startRadius) / particle->mTimeToLive;
+        
+        particle->mAngle = a;
+        particle->mDegreesPerSecond = D2R * (this->mR
```

Comment:
I think this should be `this->mDuration <= this->mElapsed`
It would probably be easier to reason about if it were written `this->mElapsed >= this->mDuration`